grpc-protocol/src/main/java/com/linecorp/armeria/server/grpc/protocol/AbstractUnsafeUnaryGrpcService.java

```diff
@@ -85,17 +124,24 @@ protected final HttpResponse doPost(ServiceRequestContext ctx, HttpRequest req)
                         return handleMessage(ctx, requestMessage);
                     }
                 }).thenApply(responseMessage -> {
+                    final HttpHeadersBuilder trailersBuilder = HttpHeaders.builder();
+                    GrpcTrailersUtil.addStatusMessageToTrailers(trailersBuilder, StatusCodes.OK, null);
+                    final HttpHeaders trailers = trailersBuilder.build();
                     final ArmeriaMessageFramer framer = new ArmeriaMessageFramer(
-                            ctx.alloc(), Integer.MAX_VALUE, false);
-                    final HttpData framed = framer.writePayload(responseMessage);
-                    final HttpHeadersBuilder trailers = HttpHeaders.builder();
-                    GrpcTrailersUtil.addStatusMessageToTrailers(trailers, StatusCodes.OK, null);
-                    return HttpResponse.of(
-                            RESPONSE_HEADERS,
-                            framed,
-                            trailers.build());
+                            ctx.alloc(), Integer.MAX_VALUE, isGrpcWebText);
+                    final HttpData content = framer.writePayload(responseMessage);
+                    final ResponseHeaders responseHeaders = RESPONSE_HEADERS_MAP.get(serializationFormat);
+                    if (UnaryGrpcSerializationFormats.isGrpcWeb(serializationFormat)) {
+                        // Send trailer as a part of the body for gRPC-web.
+                        final HttpData serializedTrailers = framer.writePayload(
+                                GrpcTrailersUtil.serializeTrailersAsMessage(ctx.alloc(), trailers), true);
+                        return HttpResponse.of(responseHeaders, content, serializedTrailers);
+                    }
+                    return HttpResponse.of(responseHeaders, content, trailers);
                 }).exceptionally(t -> {
-                    final ResponseHeadersBuilder trailers = RESPONSE_HEADERS.toBuilder();
+                    // Send Trailers-Only → HTTP-Status Content-Type Trailers.
+                    final ResponseHeadersBuilder trailers = ResponseHeaders
```

Comment:
ditto here. How about adding the trailers to the `RequestContext`?

Reply:
Discussed in https://github.com/line/armeria/pull/3716#discussion_r676405645, created https://github.com/line/armeria/issues/3724.